solana/restaking-v2/programs/restaking-v2/src/lib.rs

```diff
@@ -0,0 +1,436 @@
+
+        let set_stake_ix = solana_ibc::cpi::accounts::SetStake {
+            sender: ctx.accounts.staker.to_account_info(),
+            chain: ctx.accounts.chain.to_account_info(),
+            trie: ctx.accounts.trie.to_account_info(),
+            system_program: ctx.accounts.system_program.to_account_info(),
+            instruction: ctx.accounts.instruction.to_account_info(),
+        };
+
+        let cpi_ctx = CpiContext::new(
+            ctx.accounts.guest_chain_program.to_account_info(),
+            set_stake_ix,
+        );
+
+        let set_stake_arg = common_state
+            .validators
+            .iter()
+            .map(|validator| {
+                (
+                    sigverify::ed25519::PubKey::from(validator.clone()),
+                    stake_per_validator as i128,
+                )
+            })
+            .collect::<Vec<_>>();
+
+        solana_ibc::cpi::update_stake(cpi_ctx, set_stake_arg)?;
+
+        Ok(())
+    }
+
+    /// Withdraw tokens from the escrow and burn receipt tokens while updating the
+    /// stake for the validators on the guest chain.
+    ///
+    /// Fails if
+    /// - staker does not have enough receipt tokens to burn
+    /// - accounts needed to call guest chain program are missing
+    pub fn withdraw(ctx: Context<Withdraw>, amount: u64) -> Result<()> {
+        let common_state = &mut ctx.accounts.common_state;
+
+        let bump = ctx.bumps.common_state;
+        let seeds = [COMMON_SEED, core::slice::from_ref(&bump)];
+        let seeds = seeds.as_ref();
+        let seeds = core::slice::from_ref(&seeds);
+
+        // Check if balance is enough
+        let staker_receipt_token_account = &ctx.accounts.staker_receipt_token_account;
+
+        if staker_receipt_token_account.amount < amount {
+            return Err(error!(ErrorCodes::NotEnoughReceiptTokensToWithdraw));
+        }
+
+        let transfer_ix = Transfer {
+            from: ctx.accounts.escrow_token_account.to_account_info(),
+            to: ctx.accounts.staker_token_account.to_account_info(),
+            authority: common_state.to_account_info(),
+        };
+
+        let cpi_ctx = CpiContext::new_with_signer(
+            ctx.accounts.token_program.to_account_info(),
+            transfer_ix,
+            seeds,
+        );
+
+        anchor_spl::token::transfer(cpi_ctx, amount)?;
+
+        let burn_ix = Burn {
+            mint: ctx.accounts.receipt_token_mint.to_account_info(),
+            from: ctx.accounts.staker_receipt_token_account.to_account_info(),
+            authority: ctx.accounts.staker.to_account_info(),
+        };
+
+        let cpi_ctx = CpiContext::new(ctx.accounts.token_program.to_account_info(), burn_ix);
+
+        anchor_spl::token::burn(cpi_ctx, amount)?;
+
+        // Call guest chain program to update the stake equally
+        let stake_per_validator = (amount / common_state.validators.len() as u64) as i128;
```

Comment:
Similarly here, we may en up reducing stake by less than the amount we’ve burnt. Perhaps round down the amount?

Reply:
yes